src/go/configgenerator/routegen/routegentest/routegentest.go

```diff
@@ -0,0 +1,184 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package routegentest contains test helpers to test route generators.
+package routegentest
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/configgenerator/routegen"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
+	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
+	routepb "github.com/envoyproxy/go-control-plane/envoy/config/route/v3"
+	"github.com/imdario/mergo"
+	servicepb "google.golang.org/genproto/googleapis/api/serviceconfig"
+)
+
+// SuccessOPTestCase is the shared struct to test a RouteGenerator with
+// One Platform config. It checks the factory and generator both succeed.
+type SuccessOPTestCase struct {
+	Desc string
+
+	ServiceConfigIn *servicepb.Service
+
+	// OptsIn is the input ESPv2 Options.
+	// Will be merged with defaults.
+	OptsIn options.ConfigGeneratorOptions
+
+	// OptsMergeBehavior is implementation detail on how opts are merged into
+	// default ESPv2 opts. Useful to manually set when dealing with opts
+	// that default to true, but you want to set to false (empty value).
+	//
+	// See https://github.com/imdario/mergo/issues/129 for an example.
+	//
+	// If specified, make sure `OptsIn` contains ALL options you need, as defaults
+	// will most likely be ignored.
+	OptsMergeBehavior func(*mergo.Config)
+
+	// WantHostConfig is the expected virtual host config, in JSON.
+	// Only the routes are populated and verified against the generated routes.
+	WantHostConfig string
+}
+
+// RunTest is a test helper to run the test.
+func (tc *SuccessOPTestCase) RunTest(t *testing.T, factory routegen.RouteGeneratorOPFactory) {
+	t.Helper()
+
+	if tc.OptsMergeBehavior == nil {
+		tc.OptsMergeBehavior = mergo.WithOverride
+	}
+	t.Run(tc.Desc, func(t *testing.T) {
+		opts := options.DefaultConfigGeneratorOptions()
+		if err := mergo.Merge(&opts, tc.OptsIn, tc.OptsMergeBehavior); err != nil {
+			t.Fatalf("Merge() of test opts into default opts got err: %v", err)
+		}
+
+		gotGenerators, err := factory(tc.ServiceConfigIn, opts)
+		if err != nil {
+			t.Fatalf("NewXYZRouteGensFromOPConfig() got error: %v", err)
```

Comment:
Change XYZ to the respective name?

Reply:
Requires using reflection to get the factory function name. We shouldn't introduce reflection unnecessarily.